src/core/mainloop.rkt

```diff
@@ -237,8 +249,16 @@
 (define (run-iteration!)
   (unless (^next-alts^)
     (choose-alts!))
-  (define locs (append-map (compose all-subexpressions alt-expr) (^next-alts^)))
-  (reconstruct! (generate-candidates (remove-duplicates locs)))
+
+  (define global-batch (progs->batch (map alt-expr (^next-alts^))))
+  (define (make-batchref x idx)
+    (struct-copy alt x [expr (batchref global-batch idx)]))
+
+  (^next-alts^ (map make-batchref (^next-alts^) (vector->list (batch-roots global-batch))))
+  (define roots (batch-alive-nodes global-batch #:condition node-is-impl?))
```

Comment:
This is fine but why would we have any not-alive nodes? Like didn't we just make the batch?